src/rgw/rgw_flight_frontend.cc

```diff
@@ -0,0 +1,224 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+
+#include <cstdio>
+#include <filesystem>
+#include <sstream>
+
+#include "arrow/type.h"
+#include "arrow/flight/server.h"
+#include "arrow/io/file.h"
+
+#include "parquet/arrow/reader.h"
+#include "parquet/arrow/schema.h"
+#include "parquet/stream_reader.h"
+
+#include "rgw_flight_frontend.h"
+#include "rgw_flight.h"
+
+
+#define dout_subsys ceph_subsys_rgw_flight
+
+
+namespace rgw::flight {
+
+  const FlightKey null_flight_key = 0;
+
+  FlightFrontend::FlightFrontend(boost::intrusive_ptr<ceph::common::CephContext> _cct,
+				 RGWFrontendConfig* _config,
+				 rgw::sal::Store* _store,
+				 int _port) :
+    cct(_cct),
+    dp(cct.get(), dout_subsys, "rgw arrow_flight: "),
+    config(_config),
+    port(_port)
+  {
+    FlightStore* flight_store = new MemoryFlightStore(_cct);
+    flight_server = new FlightServer(_cct, _store, flight_store);
+  }
+
+  FlightFrontend::~FlightFrontend() {
+    delete flight_server;
+  }
+
+  int FlightFrontend::init() {
+    if (port <= 0) {
+      port = FlightServer::default_port;
+    }
+    const std::string url =
+      std::string("grpc+tcp://localhost:") + std::to_string(port);
+    flt::Location location;
+    arw::Status s = flt::Location::Parse(url, &location);
+    if (!s.ok()) {
+      ERROR << "couldn't parse url=" << url << ", status=" << s << dendl;
+      return -EINVAL;
+    }
+
+    flt::FlightServerOptions options(location);
+    options.verify_client = false;
+    s = flight_server->Init(options);
+    if (!s.ok()) {
+      ERROR << "couldn't init flight server; status=" << s << dendl;
+      return -EINVAL;
+    }
+
+    INFO << "FlightServer inited; will use port " << port << dendl;
+    return 0;
+  }
+
+  int FlightFrontend::run() {
+    try {
+      flight_thread = make_named_thread(server_thread_name,
+					&FlightServer::Serve,
+					flight_server);
+      set_flight_server(flight_server);
+
+      INFO << "FlightServer thread started, id=" <<
+	flight_thread.get_id() <<
+	", joinable=" << flight_thread.joinable() << dendl;
+      return 0;
+    } catch (std::system_error& e) {
+      ERROR << "FlightServer thread failed to start" << dendl;
+      return -ENOSPC;
```

Comment:
system_error contains an error_code, so you can return `-e.code().value()` (nit: catch as const ref)